File: src/algorithms/arrays/largest-rectangle-histogram/sources/LargestRectangleHistogram.java
// Largest Rectangle in Histogram — O(n) monotonic stack approach
import java.util.ArrayDeque;
import java.util.Deque;

public class LargestRectangleHistogram {
    public static int[] largestRectangleHistogram(int[] heights) {
        int arrayLength = heights.length;
        if (arrayLength == 0) { // @step:initialize
            return new int[]{0, -1, -1, 0}; // @step:initialize
        }

        Deque<Integer> indexStack = new ArrayDeque<>(); // @step:initialize
        int maxArea = 0; // @step:initialize
        int bestLeft = 0; // @step:initialize
        int bestRight = 0; // @step:initialize
        int bestHeight = 0; // @step:initialize

        for (int currentIndex = 0; currentIndex <= arrayLength; currentIndex++) {
            int currentHeight = currentIndex == arrayLength ? 0 : heights[currentIndex]; // @step:compare

            while (!indexStack.isEmpty() && currentHeight < heights[indexStack.peek()]) { // @step:compare
                int poppedIndex = indexStack.pop(); // @step:visit
                int poppedHeight = heights[poppedIndex]; // @step:visit
                int leftBoundary = indexStack.isEmpty() ? 0 : indexStack.peek() + 1; // @step:visit
                int width = currentIndex - leftBoundary; // @step:visit
                int area = poppedHeight * width; // @step:visit

                if (area > maxArea) { // @step:compare
                    maxArea = area; // @step:visit
                    bestLeft = leftBoundary; // @step:visit
                    bestRight = currentIndex - 1; // @step:visit
                    bestHeight = poppedHeight; // @step:visit
                }
            }

            indexStack.push(currentIndex); // @step:visit
        }

        return new int[]{maxArea, bestLeft, bestRight, bestHeight}; // @step:complete
    }
}

File: src/algorithms/arrays/stack-based/largest-rectangle-histogram/sources/largest-rectangle-histogram.py
# Largest Rectangle in Histogram — O(n) monotonic stack approach
def largest_rectangle_histogram(heights: list[int]) -> dict:
    array_length = len(heights)
    if array_length == 0:  # @step:initialize
        return {"max_area": 0, "left_index": -1, "right_index": -1, "height": 0}  # @step:initialize

    index_stack = []  # @step:initialize
    max_area = 0  # @step:initialize
    best_left = 0  # @step:initialize
    best_right = 0  # @step:initialize
    best_height = 0  # @step:initialize

    for current_index in range(array_length + 1):
        current_height = 0 if current_index == array_length else heights[current_index]  # @step:compare

        while index_stack and current_height < heights[index_stack[-1]]:  # @step:compare
            popped_index = index_stack.pop()  # @step:visit
            popped_height = heights[popped_index]  # @step:visit
            left_boundary = 0 if not index_stack else index_stack[-1] + 1  # @step:visit
            width = current_index - left_boundary  # @step:visit
            area = popped_height * width  # @step:visit

            if area > max_area:  # @step:compare
                max_area = area  # @step:visit
                best_left = left_boundary  # @step:visit
                best_right = current_index - 1  # @step:visit
                best_height = popped_height  # @step:visit

        index_stack.append(current_index)  # @step:visit

    return {"max_area": max_area, "left_index": best_left, "right_index": best_right, "height": best_height}  # @step:complete
